File: SudokuSolver.java
public class SudokuSolver {

    static int[][] board = {
        { 3, 0, 6, 5, 0, 8, 4, 0, 0 },
        { 5, 2, 0, 0, 0, 0, 0, 0, 0 },
        { 0, 8, 7, 0, 0, 0, 0, 3, 1 },
        { 0, 0, 3, 0, 1, 0, 0, 8, 0 },
        { 9, 0, 0, 8, 6, 3, 0, 0, 5 },
        { 0, 5, 0, 0, 9, 0, 6, 0, 0 },
        { 1, 3, 0, 0, 0, 0, 2, 5, 0 },
        { 0, 0, 0, 0, 0, 0, 0, 7, 4 },
        { 0, 0, 5, 2, 0, 6, 3, 0, 0 }
    };
    public static void printBoard() {
        for(int i = 0; i < board.length; i++) {
            for(int j = 0; j < board[i].length; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
        System.out.println();
        
    }

    //check cell
    public static boolean validate(int r, int c, int[][] currBoard, int num) {
        //check row
        for(int i = 0; i < currBoard[0].length; i++) {
            if(currBoard[r][i] == num) return false;
        }
        //check column
        for(int i = 0; i < currBoard.length; i++) {
            if(currBoard[i][c] == num) return false;
        }
        //check square
        int squareR = r - r%3;
        int squareC = c - c%3;

        for(int i = squareR; i < squareR+3; i++) {
            for(int j = squareC; j < squareC+3; j++) {
                if(currBoard[i][j] == num) return false;
            }
        }

        return true;
    }

    public static boolean solve(int[][] board) {
        //check if board is done
        boolean done = true;
        int r = -1;
        int c = -1;
        for(int i = 0; i < board.length; i++) {
            for(int j = 0; j < board[0].length; j++) {
                if(board[i][j] == 0) {
                    done = false;
                    r = i;
                    c = j;
                    break;
                }
            }
            if(!done) break;
        }

        //if board is done return true
        if(done) return true;

        //if not done try all numbers through backtracking
        for(int i = 1; i < 10; i++) {
            if(validate(r, c, board, i)) {
                board[r][c] = i;
                if(solve(board)) {
                    return true;
                } else {
                    board[r][c] = 0;
                }
            }
        } 
        return false;


    }
    public static void main(String[] args) {
        if(solve(board)) {
            printBoard();
        } else {
            System.out.println("not solvable");
        }
    }
}

File: SudokuSolver.js
import { Util } from "./Util.js";


function checkNum(board, r, c, num) {
    //check row for same num
    for(let col = 0; col < 9; col++) {
        if(board[r][col] === num) {
            return false;
        }
    }

    //check col for same num
    for(let row = 0; row < 9; row++) {
        if(board[row][c] === num) {
            return false;
        }
    }

    //check square for same num
    for(let i = r-(r%3); i < r-(r%3)+3; i++) {
        for(let j = c-(c%3); j < c-(c%3)+3; j++) {
            if(i != r && j != c) {
                if(board[i][j] === num) {
                    return false;
                }
            }
        }
    }

    return true;
}

function checkValid(board) {
    for(let i = 0; i < 9; i++) {
        for(let j = 0; j < 9; j++) {
            if(!checkNum(board, i, j, board[i][j])) {
                console.log({i, j})
                return false;
            }
        }
    }
    return true;
}

function isComplete(board) {
    for(let i = 0; i < 9; i++) {
        for(let j = 0; j < 9; j++) {
            if(board[i][j] === 0) {
                return false;
            }
        }
    }
    return true;
}

function adjNumRecur(board, r, c, num) {
    if(num === 10) {
        return 0;
    }
    if(!checkNum(board, r, c, num)) {
        return adjNumRecur(board, r, c, num+1);
    } else {
        return num;
    }
}

function shuffle(array) {
    let currIdx = array.length, randIdx;

    while(currIdx > 0) {
        randIdx = Math.floor(Math.random() * currIdx);
        currIdx--;

        [array[currIdx], array[randIdx]] = [array[randIdx], array[currIdx]];
    }

    return array;
}

function solve(board) {
    const initialize2DArray = (w, h, val = null) =>
        Array.from({ length:h }).map(() => Array.from({ length: w }).fill(val));
    var isBlank = initialize2DArray(9, 9, false);
    var r = 0;
    var c = 0;
    var firstR = -1;
    var firstC = -1; 
    var firstPass = true;

    while(r < 9 && c < 9) {
        if(board[r][c] != 0 && isBlank[r][c] == false) {
            if(c == 8) {
                c = 0;
                r++;
            } else {
                c++;
            }
            continue;
        } else {
            isBlank[r][c] = true;
            if(firstPass) {
                firstC = c;
                firstR = r;
                firstPass = false;
            }
        }

        board[r][c] = adjNumRecur(board, r, c, board[r][c]+1);
        if(board[r][c] == 0) {
            if(firstC == c && firstR == r) {
                console.log("No solution exists");
                return;
            }
            if(c == 0) {
                c = 8;
                r--;
            } else {
                c--;
            }  
            while(isBlank[r][c] != true) {
                if(c == 0) {
                    c = 8;
                    r--;
                } else {
                    c--;
                }        
            }
        } else {
            if(c == 8) {
                c = 0;
                r++;
            } else {
                c++;
            }
        }
    }
}

function createSudoku() {
    let sudoku = getRandomSudoku();

    solve(sudoku);

    for(let i = 0; i < 9; i++) {
        for(let j = 0; j < 9; j++) {
            if(Math.random() > 0.3) {
                sudoku[i][j] = 0;
            }
        }
    }
}

function getRandomSudoku() {
    const initialize2DArray = (w, h, val = null) =>
        Array.from({ length:h }).map(() => Array.from({ length: w }).fill(val));
    
    let sudoku = initialize2DArray(9, 9, 0);
    let solution = initialize2DArray(9, 9, 0);
    
    while(!isComplete(solution)) {
        let numDone = 0;
        sudoku = initialize2DArray(9, 9, 0);
        while(numDone < 30) {
            let row = Math.floor(Math.random() * 9);
            let col = Math.floor(Math.random() * 9);
            let num = Math.floor(Math.random() * 9) + 1;

            if(sudoku[row][col] === 0) {
                if(!checkNum(sudoku, row, col, num)) {
                    sudoku[row][col] = 0;
                } else {
                    sudoku[row][col] = num;
                    numDone++;
                }
            }
        }
        Util.copyGrid(sudoku, solution);
        solve(solution);
        //Util.print2DArray(solution);
    }
    Util.print2DArray(sudoku);
    return sudoku;
}


let solution = [];
let grid = getRandomSudoku();
Util.print2DArray(grid);
Util.copyGrid(grid, solution);
solve(solution);
Util.print2DArray(solution);